processing/src/main/java/io/druid/query/groupby/strategy/GroupByStrategyV2.java

```diff
@@ -167,11 +176,36 @@ private static int countRequiredMergeBufferNum(Query query, int foundNum)
     // until the outer groupBy processing completes.
     // This is same for subsequent groupBy layers, and thus the maximum number of required merge buffers becomes 2.
 
-    final DataSource dataSource = query.getDataSource();
-    if (foundNum == MAX_MERGE_BUFFER_NUM + 1 || !(dataSource instanceof QueryDataSource)) {
-      return foundNum - 1;
+    return countGroupByLayers(query, foundNum) - 1;
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <T> int countGroupByLayers(final Query<T> query, final int foundNum)
+  {
+    if (foundNum == MAX_MERGE_BUFFER_NUM + 1) {
+      return foundNum;
+    }
+
+    int maxFoundFromChildren = -1;
+    for (DataSourceWithSegmentSpec eachSpec : query.getDataSources()) {
+      final DataSource dataSource = eachSpec.getDataSource();
+
+      if (dataSource instanceof QueryDataSource) {
+        QueryDataSource queryDataSource = (QueryDataSource) dataSource;
+        if (queryDataSource.getQuery() instanceof GroupByQuery) {
+          final int tmp = countGroupByLayers(queryDataSource.getQuery(), foundNum + 1);
+          maxFoundFromChildren = tmp > maxFoundFromChildren ? tmp : maxFoundFromChildren;
+        } else {
+          final int tmp = countGroupByLayers(queryDataSource.getQuery(), 0);
```

Comment:
This method find the maximum number of consecutive groupBys. The foundNum is set to 0 because a non groupBy is found. I added a comment.